src/lib/provable/gadgets/sha256.ts

```diff
@@ -88,80 +88,22 @@ const SHA256 = {
     // padding the message $5.1.1 into blocks that are a multiple of 512
     let messageBlocks = padding(data);
 
-    const H = SHA256Constants.H.map((x) => UInt32.from(x));
+    let H = SHA256Constants.H.map((x) => UInt32.from(x));
     const K = SHA256Constants.K.map((x) => UInt32.from(x));
 
     const N = messageBlocks.length;
 
     for (let i = 0; i < N; i++) {
-      const M = messageBlocks[i];
-      // for each message block of 16 x 32bit do:
-      const W: UInt32[] = [];
-
-      // prepare message block
-      for (let t = 0; t <= 15; t++) W[t] = M[t];
-      for (let t = 16; t <= 63; t++) {
-        // the field element is unreduced and not proven to be 32bit, we will do this later to save constraints
-        let unreduced = DeltaOne(W[t - 2])
-          .value.add(W[t - 7].value)
-          .add(DeltaZero(W[t - 15]).value.add(W[t - 16].value));
-
-        // mod 32bit the unreduced field element
-        W[t] = UInt32.Unsafe.fromField(divMod32(unreduced, 16).remainder);
-      }
-
-      // initialize working variables
-      let a = H[0];
-      let b = H[1];
-      let c = H[2];
-      let d = H[3];
-      let e = H[4];
-      let f = H[5];
-      let g = H[6];
-      let h = H[7];
-
-      // main loop
-      for (let t = 0; t <= 63; t++) {
-        // T1 is unreduced and not proven to be 32bit, we will do this later to save constraints
-        const unreducedT1 = h.value
-          .add(SigmaOne(e).value)
-          .add(Ch(e, f, g).value)
-          .add(K[t].value)
-          .add(W[t].value)
-          .seal();
-
-        // T2 is also unreduced
-        const unreducedT2 = SigmaZero(a).value.add(Maj(a, b, c).value);
-
-        h = g;
-        g = f;
-        f = e;
-        e = UInt32.Unsafe.fromField(
-          divMod32(d.value.add(unreducedT1), 16).remainder
-        ); // mod 32bit the unreduced field element
-        d = c;
-        c = b;
-        b = a;
-        a = UInt32.Unsafe.fromField(
-          divMod32(unreducedT2.add(unreducedT1), 16).remainder
-        ); // mod 32bit
-      }
-
-      // new intermediate hash value
-      H[0] = H[0].addMod32(a);
-      H[1] = H[1].addMod32(b);
-      H[2] = H[2].addMod32(c);
-      H[3] = H[3].addMod32(d);
-      H[4] = H[4].addMod32(e);
-      H[5] = H[5].addMod32(f);
-      H[6] = H[6].addMod32(g);
-      H[7] = H[7].addMod32(h);
+      const W = computeMessageSchedule(messageBlocks[i]);
+      H = sha256Compression(H, W, K);
```

Comment:
see comment above; do we have to K and H  in here as well? Why not get these constants in the method directly?

Reply:
Yes - we have `K` always which will be used internally but `H` only in the case of being an `initialState`

Reply:
oh thats right, yep!